modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java

```diff
@@ -977,25 +977,27 @@ public void removeWatch(WatchListener listener) {
     public void compact(long revision) {
         assert revision >= 0;
 
-        rwLock.writeLock().lock();
+        try (RocksIterator iterator = index.newIterator()) {
+            iterator.seekToFirst();
 
-        try (WriteBatch batch = new WriteBatch()) {
-            assert revision < rev : String.format(
-                    "Compaction revision should be less than the current: [compaction=%s, current=%s]",
-                    revision, rev
-            );
+            RocksUtils.forEach(iterator, (key, value) -> {
+                rwLock.writeLock().lock();
```

Comment:
I'm a little concerned that we take a lock for every key, maybe we should batch it?

Reply:
Yes, why not.